File: netcdf/netcdf_ex1.f90
 ! * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 ! Copyright by The HDF Group.                                               *
 ! Copyright by the Board of Trustees of the University of Illinois.         *
 ! All rights reserved.                                                      *
 !                                                                           *
 ! This file is part of HDF5.  The full HDF5 copyright notice, including     *
 ! terms governing use, modification, and redistribution, is contained in    *
 ! the files COPYING and Copyright.html.  COPYING can be found at the root   *
 ! of the source code distribution tree; Copyright.html can be found at      *
 ! http://hdfgroup.org/HDF5/doc/Copyright.html.  If you do not have          *
 ! access to either file, you may request a copy from help@hdfgroup.org.     *
 ! * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

 ! Programmer:  Isayah Reed <ireed@hdfgroup.org>
 !              May  4, 2011
 !
 ! Purpose:
 !  This is a simple netCDF4 program that demonstrates how to add CF attributes
 !  to an h5 file. Creates a file with 3 variables: lat, lon, temp. Lat contains
 !  the CF attributes: units, long_name, and standard_name. Lon has the same CF
 !  attributes as the latitude variable. Temp contains the CF attributes: units,
 !  long_name, _FillValue, and coordinates. Outputs data to netcdf_ex1.h5
 !  
 ! Compile:
 !  ifort netcdf_ex1.f90 -I$(NETCDF4_DIR)/include -I$(HDF5_DIR)/include -L$(NETCDF4_DIR)/lib -L$(HDF5_DIR)/lib $(NETCDF4_DIR)/lib/libnetcdff.a $(NETCDF4_DIR)/lib/libnetcdf.a  $(HDF5_DIR)/lib/libhdf5_hl.a $(HDF5_DIR)/lib/libhdf5.a -lsz -lz -lm
 !  


PROGRAM main

  USE NETCDF

  IMPLICIT NONE

  CHARACTER(LEN=13), PARAMETER :: filename = "netcdf_ex1.nc"
  CHARACTER(LEN=4) , PARAMETER :: TEMP= "temp"
  CHARACTER(LEN=3) , PARAMETER :: LAT= "lat"
  CHARACTER(LEN=3) , PARAMETER :: LON= "lon"
  CHARACTER(LEN=9) , PARAMETER :: LONGNAME = "long_name"
  CHARACTER(LEN=5) , PARAMETER :: UNITS = "units"
  CHARACTER(LEN=10) , PARAMETER :: FILLVALUE = "_FillValue"
  CHARACTER(LEN=11) , PARAMETER :: COORDINATES = "coordinates"

  INTEGER          , PARAMETER :: nx1     = 180
  INTEGER          , PARAMETER :: ny1     = 360

  INTEGER :: status
  
  INTEGER :: ncid, varid0, varid1, varid2 ! handles
  CHARACTER(LEN=20), DIMENSION(2) ::  attr_data  ! Attribute string data
  ! switch dimensions to adhere to C standards
  INTEGER, DIMENSION(1:2) :: temp_dims = (/ny1, nx1/) ! size read/write buffer
  INTEGER :: lat_dims != (/nx1/) 
  INTEGER :: lon_dims != (/ny1/) 
  INTEGER, DIMENSION(1) :: dims1
  INTEGER, DIMENSION(2) :: dims2
  DOUBLE PRECISION, DIMENSION(1:nx1,1:ny1) :: temp_data
  DOUBLE PRECISION, DIMENSION(nx1) :: lat_data
  DOUBLE PRECISION, DIMENSION(ny1) :: lon_data
  DOUBLE PRECISION, DIMENSION(1) :: fill_data
  INTEGER :: i, j, strlen
  DOUBLE PRECISION :: fill_value= -999.0
  CHARACTER, DIMENSION(2) ::  hvl_data

  !
  ! Initialize FORTRAN interface.
  !
  ! CALL h5open_f(hdferr)
  !
  ! Initialize temperature data.
  !
  DO i = 1, 60
     DO j = 1, ny1 
        temp_data(i,j) = 280.0
     ENDDO
  ENDDO
  DO i = 61, 120
     DO j = 1, ny1 
        temp_data(i,j) = 300.0
     ENDDO
  ENDDO
  DO i = 121, nx1
     DO j = 1, ny1 
        temp_data(i,j) = 280.0
     ENDDO
  ENDDO


  lat_data(1)= -90.0
  DO i = 2, nx1
    lat_data(i) = lat_data(i-1)+1.0
  ENDDO


  lon_data(1)= -180.0
  DO i = 2, ny1
    lon_data(i) = lon_data(i-1)+1.0
  ENDDO

  !
  ! Create a new file using the default properties.
  ! 
  status=  nf90_create(path=filename, cmode=IOR(NF90_CLOBBER,NF90_HDF5), ncid=ncid)

  ! define dimensions
  status= nf90_def_dim(ncid, LAT, NX1, lat_dims)
  status= nf90_def_dim(ncid, LON, NY1, lon_dims)

  ! Define the variable. The type of the variable in this case is
  ! NC_FLOAT
  status= nf90_def_var(ncid, TEMP, NF90_FLOAT, (/lat_dims, lon_dims/), varid0)
  status= nf90_def_var(ncid, LAT, NF90_FLOAT, lat_dims, varid1)
  status= nf90_def_var(ncid, LON, NF90_FLOAT, lon_dims, varid2)

  ! define fillvalue, set no_fill=0 to write fillvalue
  status= nf90_def_var_fill(ncid, varid0, 0, -999)

  ! End define mode. This tells netCDF we are done defining metadata.
  status= nf90_enddef(ncid)


  ! don't need to adjust array for fortran-C compatibility
  status= nf90_put_var(ncid, varid0, temp_data) 
  status= nf90_put_var(ncid, varid1, lat_data) 
  status= nf90_put_var(ncid, varid2, lon_data) 

  status= nf90_put_att(ncid, varid0, UNITS, "kelvin")
  status= nf90_put_att(ncid, varid1, UNITS, "degrees_north")
  status= nf90_put_att(ncid, varid2, UNITS, "degrees_east")
  
  status= nf90_put_att(ncid, varid0, LONGNAME, "temperature")
  status= nf90_put_att(ncid, varid1, LONGNAME, "latitude")
  status= nf90_put_att(ncid, varid2, LONGNAME, "longitude")

  status= nf90_put_att(ncid, varid0, FILLVALUE, -999.0)

  ! COORDINATES
  status= nf90_put_att(ncid, varid0, COORDINATES, LAT)
  status= nf90_put_att(ncid, varid0, COORDINATES, LON)

  status= nf90_close(ncid)
  
END PROGRAM main

File: netcdf/netcdf_ex1.c

/*
 * Programmer:  Isayah Reed <help@hdfgroup.org> 
 *              May  4, 2011
 *
 * Purpose:
 *  This is a simple netCDF4 program that demonstrates how to add CF attributes
 *  to an h5 file. Creates a file with 3 variables: lat, lon, temp. Lat contains
 *  the CF attributes: units, long_name, and standard_name. Lon has the same CF
 *  attributes as the latitude variable. Temp contains the CF attributes: units,
 *  long_name, _FillValue, and coordinates. Outputs data to netcdf_ex1.h5
 *
 * Compile:
 *  gcc netcdf_ex1.c -I$NETCDF4_DIR/include -I$HDF5_DIR/include -L$NETCDF4_DIR/lib -L$HDF5_DIR/lib $NETCDF4_DIR/lib/libnetcdf.a $HDF5_DIR/lib/libhdf5_hl.a $HDF5_DIR/lib/libhdf5.a -lsz -lz -lm
 *
*/


    #include <stdlib.h>
    #include <stdio.h>
    #include "netcdf.h"
    #include "hdf5.h"
    #include "hdf5.h"
    #include "string.h"
     
     /* This is the name of the data file we will create. */
    #define FILE_NAME "netcdf_ex1.nc"
     
     /* We are writing 2D data, a 180 x 360 grid. */
    #define NDIMS 2
    #define NX1     180      /* dataset dimensions */
    #define NY1     360
    #define STRINGLISTSIZE  2
    /* datasets */
    #define TEMP "temp"
    #define LAT "lat"
    #define LON "lon"
    /* attributes */
    #define UNITS "units"
    #define FILLVALUE "_FillValue"
    #define LONGNAME "long_name"
    #define COORDINATES "coordinates"
     
     /* Handle errors by printing an error message and exiting with a
      * non-zero status. */
     #define ERRCODE 2
     #define ERR(e) {printf("Error: %s\n", nc_strerror(e)); exit(ERRCODE);}

     int main()
     {
        int ncid, varid[3];
        int dimsf[1], dimsa[2], dimst; /* float, array, text dimensions */
        float temp_array[NX1][NY1], lat_array[NX1], 
                lon_array[NY1], fillvalue= -999.0;
        int retval, i, j;
        char  *degrees_east= "degrees_east", *degrees_north= "degrees_north",
            *kelvin= "kelvin", *latitude= "latitude",
            *longitude= "longitude", *temperature= "temperature";
        char *coorlist[STRINGLISTSIZE]= {"lat\0", "lon\0"};        
     
        /* values between a[60][*] and a[120][*] is around 300.0 */
        for(i=60; i<=120; i++)
            for(j=0; j<NY1; j++)
                temp_array[i][j]= 300.0;

       /* values between a[0][*] and a[59][*], a[121][*] and a[179][*] 
            is around 280.0 */
        for(i=0; i<60; i++)
            for(j=0; j<NY1; j++)
                temp_array[i][j]= 280.0;

        for(i=121; i<NX1; i++)
            for(j=0; j<NY1; j++)
                temp_array[i][j]= 280.0;

        lat_array[0]= -90.0;
        for(i=1; i<NX1; i++)
            lat_array[i]= lat_array[i-1]+1.0;

        lon_array[0]= -180;
        for(i=1; i<NY1; i++)
            lon_array[i]= lon_array[i-1]+1.0;
     
        /* Always check the return code of every netCDF function call. In
         * this example program, any retval which is not equal to NC_NOERR
         * (0) will cause the program to print an error message and exit
         * with a non-zero return code. */
     
        /* Create the file. The NC_CLOBBER parameter tells netCDF to
         * overwrite this file, if it already exists.*/
        if ((retval = nc_create(FILE_NAME, NC_NETCDF4|NC_CLOBBER, &ncid)))
           ERR(retval);
     


        /* Define the dimensions. NetCDF will hand back an ID for each. */
        if ((retval = nc_def_dim(ncid, LAT, NX1, &dimsa[0])))
           ERR(retval);
        if ((retval = nc_def_dim(ncid, LON, NY1, &dimsa[1])))
           ERR(retval);
     
     
        /* Define the variable. The type of the variable in this case is
         * NC_INT (4-byte integer). */
        if ((retval = nc_def_var(ncid, TEMP, NC_FLOAT, 2,
                    dimsa, &varid[0])))
           ERR(retval);
        if ((retval = nc_def_var(ncid, LAT, NC_FLOAT, 1,
                    &dimsa[0], &varid[1])))
           ERR(retval);
        if ((retval = nc_def_var(ncid, LON, NC_FLOAT, 1,
                    &dimsa[1], &varid[2])))
           ERR(retval);

     
        /* define fillvalue, set no_fill=0 to write fillvalue */
        if ((retval= nc_def_var_fill(ncid, varid[0], 0, &fillvalue)))
            ERR(retval);


        /* End define mode. This tells netCDF we are done defining
         * metadata. */
        if ((retval = nc_enddef(ncid)))
           ERR(retval);


        /* Write the coordinate variable data. This will put the latitudes
           and longitudes into the netCDF file. */
        if ((retval = nc_put_var_float(ncid, varid[0], 
                &temp_array[0][0])))
           ERR(retval);

        if ((retval = nc_put_var_float(ncid, varid[1], &lat_array[0])))
           ERR(retval);

        if ((retval = nc_put_var_float(ncid, varid[2], &lon_array[0])))
           ERR(retval);

        if ((retval = nc_put_att_text(ncid, varid[0], UNITS, strlen(kelvin),
                kelvin)))
            ERR(retval);

        if ((retval = nc_put_att_text(ncid, varid[1], UNITS, 
                strlen(degrees_north), degrees_north)))
            ERR(retval);

        if ((retval = nc_put_att_text(ncid, varid[2], UNITS, 
                strlen(degrees_east), degrees_east)))
            ERR(retval);
     
        if ((retval = nc_put_att_text(ncid, varid[0], LONGNAME, 
                strlen(temperature), temperature)))
            ERR(retval);

        if ((retval = nc_put_att_text(ncid, varid[1], LONGNAME, 
                strlen(latitude), latitude)))
            ERR(retval);

        if ((retval = nc_put_att_text(ncid, varid[2], LONGNAME, 
                strlen(longitude), longitude)))
            ERR(retval);

        /* use nc_put_att_string for string lists */
        if ((retval = nc_put_att_string(ncid, varid[0], COORDINATES, 
                STRINGLISTSIZE, (const char**)&coorlist)))
            ERR(retval);

        /* Close the file. This frees up any internal netCDF resources
         * associated with the file, and flushes any buffers. */
        if ((retval = nc_close(ncid)))
           ERR(retval);
     
        return 0;
     }

